Add upload data styles to Splunk uploader

Tatsuya-hasegawa opened the pull request:
Hi

I have added the post styles to Splunk by Splunk Uploader.
It's very useful for msticpy Splunk uploader's users to post data by json or text style.

ASIS, all the data frame are posted with the csv style by using to_csv() method.
It occurs event fragments by depending on the data to upload.

So, I developed a switch by using "sourcetype" argument. The main change is below. 

```python
__post_data
        for row in data.iterrows():
            if sourcetype=="csv":
                data = row[1].to_csv()  # type: ignore
            elif sourcetype=="json":
                data = row[1].to_json()  # type: ignore
            else:
                data = row[1].to_string()  # type: ignore
```

In addition, I renamed table_name to sourcetype at many locations.
Because, table_name is not usual word in Splunk.
On the other hand, pylint was failed in upload_file and upload_folder methods due to W0237 so I remain table_name in only these methods.

pylint...................................................................Failed
msticpy/data/uploaders/splunk_uploader.py:166:4: W0237: Parameter 'table_name' has been renamed to 'sourcetype' in overriding 'SplunkUploader.upload_file' method (arguments-renamed)
msticpy/data/uploaders/splunk_uploader.py:218:4: W0237: Parameter 'table_name' has been renamed to 'sourcetype' in overriding 'SplunkUploader.upload_folder' method (arguments-renamed)

Finally, my code commit has been passed by the commit test.

```shell
(base) hacket@hackeTlab msticpy % git commit -m "add_post_data_styles_to_splunk_uploader"
Check Yaml...........................................(no files to check)Skipped
Check JSON...........................................(no files to check)Skipped
Trim Trailing Whitespace.................................................Passed
black....................................................................Passed
pylint...................................................................Passed
flake8...................................................................Passed
isort (python)...........................................................Passed
pydocstyle...............................................................Passed
check_reqs_all...........................................................Passed
[mod_splunk_uploader_post 10646520] add_post_data_styles_to_splunk_uploader
 2 files changed, 41 insertions(+), 18 deletions(-)
 ```
 
 Best regards,

Tatsuya-hasegawa commented:
My Splunk upload test is like this.

![image](https://github.com/microsoft/msticpy/assets/40039738/8864b36d-dc1f-489a-9871-738d90155f16)

## csv example for the test data is bad!
![image](https://github.com/microsoft/msticpy/assets/40039738/9bb20435-377c-410f-8cbc-332f66397ea0)


## json example for the test data is great!
![image](https://github.com/microsoft/msticpy/assets/40039738/5652727c-4973-483c-8a19-3769334d1052)

## text example for the test data is not bad!
![image](https://github.com/microsoft/msticpy/assets/40039738/3123be42-481a-4047-82a7-02351bd6cfb9)

Thanks in advance.

ianhelle commented on msticpy/data/uploaders/splunk_uploader.py:
```diff
@@ -199,7 +210,7 @@ def upload_file(  # type: ignore
         self._post_data(
             data=data,
-            table_name=table_name,
+            sourcetype=table_name,
```
Should this be sourcetype? like the other functions?

ianhelle commented on msticpy/data/uploaders/splunk_uploader.py:
```diff
@@ -116,7 +123,7 @@ def _post_data(
         self,
         data: pd.DataFrame,
-        table_name: Optional[str],
+        sourcetype: Optional[str],
```
given that you are changing the parameter name, I would try to keep backward compatibility and maybe opt for json as the default. I have a mild preference for snake-case compound words like this - source_type vs sourcetype. I'm not super concerned about that though
```python

    source_type = kwargs.get("source_type", kwargs.get("table_name", "json"))
```

ianhelle commented on msticpy/data/uploaders/splunk_uploader.py:
```diff
@@ -252,14 +265,19 @@ def upload_folder(  # type: ignore
             )
             f_progress.update(1)
             if self._debug is True:
-                print(f"{str(path)} uploaded to {table_name}")
```
probably get rid of references to table name in the body of the function - if it isn't actually the table_name, it's bit confusing.

ianhelle reviewed (changes requested):
Looks great - had a couple of minor comments..

ianhelle commented on docs/source/data_acquisition/UploadData.rst:
```diff
@@ -104,24 +104,27 @@ Uploading a DataFrame to Splunk
-value you can pass this with the ``delim`` parameter. You can specify a table name to upload the data to with that ``table_name``
-parameter but by default the uploader will upload to a table with the same name as the file. As with uploading a DataFrame
+want the data uploaded to. By default a comma separated value file is expected but if your file has some other separator
+value you can pass this with the ``delim`` parameter. You can specify the sourcetype to upload the data to with that ``table_name`` parameter but by default the uploader will upload to the sourcetype with the same name as the file.
```
Trye to keep lines wrapped to about 80-90 chars

ianhelle commented:
Looks like you need to fix a couple of tests as well because of the parameter change. You can add tests that use both table_name and source_type to check that they both work.

Tatsuya-hasegawa commented:
Hi @ianhelle 

Thanks for your check. I agree with your comments.

I wanted to rename the parameter of table_name to source_type(sourcetype).
Unfortunately, pylint W0237 prevented it.

pylint...................................................................Failed
msticpy/data/uploaders/splunk_uploader.py:166:4: W0237: Parameter 'table_name' has been renamed to 'sourcetype' in overriding 'SplunkUploader.upload_file' method (arguments-renamed)
msticpy/data/uploaders/splunk_uploader.py:218:4: W0237: Parameter 'table_name' has been renamed to 'sourcetype' in overriding 'SplunkUploader.upload_folder' method (arguments-renamed)

I don't know the way of bypass this check. Would you teach me ?
In addition, regarding to your last comment of 
> You can add tests that use both table_name and source_type to check that they both work.
How do I write that test ? Don't you mind if you give me a sample ?

Sincerely,

Tatsuya-hasegawa commented:
Dear @ianhelle 

Thanks for looking at my PR.
I have fixed codes to be along your comment, in addition,  added the tests.

Could you check again ?

Best regards,

ianhelle commented on msticpy/data/uploaders/splunk_uploader.py:
```diff
@@ -97,27 +99,35 @@ def _post_data(
     # pylint: disable=arguments-differ
     def upload_df(  # type: ignore
         self,
-        data: pd.DataFrame,
```
Would be good not to change the order of existing parameters - this is what is breaking tests.
I always add new parameters to the end of the existing named parameters  - that way it won't break existing cases that are using positional parameters.

Tatsuya-hasegawa replied:
Thanks very much for your advice!

Tatsuya-hasegawa commented:
Dear @ianhelle 

Thank you for teaching me many times. 
I have modified along with your advice.
I much appreciate your advice. 

So, I think this will finally passed for CI workflows.
My local code review is all green and my Splunk instance upload tests were also fine!

```
(base) hacket@hackeTlab msticpy % git commit -m "fix the new argument position and slightly modified index_name param default value"
Check Yaml...........................................(no files to check)Skipped
Check JSON...........................................(no files to check)Skipped
Trim Trailing Whitespace.................................................Passed
black....................................................................Passed
pylint...................................................................Passed
flake8...................................................................Passed
isort (python)...........................................................Passed
pydocstyle...............................................................Passed
check_reqs_all...........................................................Passed
[mod_splunk_uploader_post c40ed2e7] fix the new argument position and slightly modified index_name param default value
 1 file changed, 12 insertions(+), 6 deletions(-)
``` 

Best regards,

Tatsuya-hasegawa commented:
Thanks for your support to this merge.
